python/src/nanoarrow/iterator.py

```diff
@@ -276,6 +301,217 @@ def _iter1(self, offset, length):
+            if next(child_iter) is False:
+                return False
+
+        return self._out.write("}")
+
+    def _struct_iter(self, offset, length):
+        offset += self._array_view.offset
+        for i in range(length):
+            yield self._write_struct_item(offset + i)
+
+    def _string_iter(self, offset, length):
+        # A variant of iterating over strings that ensures that very large
+        # elements are not fully materialized as a Python object.
+        memoryviews = super()._binary_view_iter(offset, length)
+
+        # In the end-member scenario where every code point is four bytes,
+        # ensure we still slice enough bytes to fill max_width. Give some
+        # bytes on the end in case the last byte is the beginning of a
+        # multibyte character and to ensure we never get the trailing quote
+        # for an incomplete repr
+        max_width_slice_bytes = (self._out._max_size + 2) * 4
+        for mv in memoryviews:
+            if mv is None:
+                yield self._out.write_null()
+            else:
+                str_begin = bytes(mv[:max_width_slice_bytes]).decode()
+                yield self._out.write(repr(str_begin))
+
+    def _binary_iter(self, offset, length):
+        # A variant of iterating over strings that ensures that very large
+        # elements are not fully materialized as a Python object.
+
+        memoryviews = super()._binary_view_iter(offset, length)
+        # Give some extra bytes to ensure we never get a trailing ' for an
+        # incomplete repr.
+        max_width_slice_bytes = self._out._max_size + 4
+        for mv in memoryviews:
+            if mv is None:
+                yield self._out.write_null()
+            else:
+                yield self._out.write(repr(bytes(mv[:max_width_slice_bytes])))
+
+    def _primitive_iter(self, offset, length):
+        # These types can have relatively long reprs (e.g., large int64) but they
+        # are never so large that they can hang the repr (with the possible
+        # exception of fixed-size binary in rare cases).
+        for item in super()._primitive_iter(offset, length):
+            if item is None:
+                yield self._out.write_null()
+            else:
+                yield self._out.write(repr(item))
+
+
+class ReprBuilder:
+    """Stateful string builder for building repr strings
+
+    A wrapper around io.StringIO() that keeps track of how many
+    characters have been written vs some maximum size.
+    """
+
+    def __init__(self, max_size) -> None:
+        self._out = StringIO()
+        self._max_size = max_size
+        self._size = 0
+
+    def full(self):
+        return self._size > self._max_size
+
+    def write_null(self):
+        return self.write(repr(None))
+
+    def write(self, content):
+        """Write string content. Returns True if there are not yet max_size
+        characters in this items's repr and False otherwise."""
+        self._out.write(content)
```

Comment:
What if instead we only write the first bytes of content equal to `self._max_size - self._size`? This could get rid of unnecessary writing of data that would otherwise be truncated in `finish()`